### DISCUSSION TITLE
How to setup the entire immich service on an external drive.

### DISCUSSION THREAD

**Discussion by pbhereandnow** [immich-app/immich#10196] [Q&A]
Hi,
I wish to keep the entire immich container, database, data, DCIM, etc, EVERYTHING on an external SSD. I have set the .env accordingly. PFA the screenshot of the same.
![image](https://github.com/immich-app/immich/assets/67541372/be3754fb-9a2b-46c1-b9e8-ab0bb3152552)

On starting up the service I am getting the following error in the "immich_postgres" service. PFA the exact error.
immich_postgres          | The database cluster will be initialized with locale "en_US.utf8".
immich_postgres          | The default database encoding has accordingly been set to "UTF8".
immich_postgres          | The default text search configuration will be set to "english".
immich_postgres          | 
immich_postgres          | Data page checksums are enabled.
immich_postgres          | 
immich_postgres          | initdb: error: could not change permissions of directory "/var/lib/postgresql/data": Operation not permitted
immich_postgres          | fixing permissions on existing directory /var/lib/postgresql/data ... chmod: changing permissions of '/var/lib/postgresql/data': Operation not permitted
immich_postgres          | The files belonging to this database system will be owned by user "postgres".
immich_postgres          | This user must also own the server process.
immich_postgres          | 
immich_postgres          | The database cluster will be initialized with locale "en_US.utf8".
immich_postgres          | The default database encoding has accordingly been set to "UTF8".
immich_postgres          | The default text search configuration will be set to "english".

Kindly suggest how to resolve this error.

**Comment by bo0tzz** [comment]
What is the filesystem on the drive? Postgres won't work on filesystems like NTFS or FAT.

**Reply by pbhereandnow** [reply]
The filesystem is exFAT. I am using Apple photos to import photos and videos from my iPhone to a iPhoto library. Using the source/originals folder inside this library, I am mounting the folder as an external library. As Immich builds the database, the size of the container blows up. I wish to keep the entire thing on my external SSD.

**Comment by scaryspace** [comment]
Got the exact same query. Using Ubuntu server, with the other drive being a network share that is mounted. Have checked and the share has root perms, but cant figure out why I'm getting this issue.

**Reply by julienR2** [reply]
A bit late in the game, but I'm notw facing the same issue, and from what I've found, ExFat format doesn't support changing ownership or permissions. This explains why you get `fixing permissions on existing directory /var/lib/postgresql/data ... chmod: changing permissions of '/var/lib/postgresql/data': Operation not permitted`.

From what I've seen, it's recommended to format your hard drive to ext4 if you want to use it as the main storage for Immich upload and postgres folders.

Hope this helps

**Comment by dvishal485** [comment]
Hey, I'm using Docker compose method to host (on Windows with Docker Desktop using wsl2). I'm a bit confused about disk storage being mounted.

I have a default internal SSD (mounted as `C:` drive), if I set the path on this drive, say

```env
UPLOAD_LOCATION="C:\\Users\\<user>\\Downloads\\immich\\library"
DB_DATA_LOCATION="C:\\Users\\<user>\\Downloads\\immich\\postgres"
```

**Immich works exactly as anyone could expect it to when using `C:` drive.** Restarting container / host system doesn't have any ill-impact on its usage.

Things gets messy when I use my external HDD (mounted at `D:`)

```env
UPLOAD_LOCATION="D:\\immich\\library"
DB_DATA_LOCATION="D:\\immich\\postgres"
```

Immich works fine on **first startup**, but after I restart my device (or say stop and restart the docker containers), the first time setup reappears.
I tried to understand the issue, it turns out if I copy the directories to `C:` drive and re-run container with path updated to `C:` drive, it brings me up to the login page, and I can use Immich normally.

```
Copy D:\\immich\\library -> C:\\Users\\<user>\\Downloads\\immich\\library
Copy D:\\immich\\postgres-> C:\\Users\\<user>\\Downloads\\immich\\postgres
```

Another thing which I tried is keeping the `library` on `D:` drive, and `postgres` on `C:` drive. (which although is not recommended by Docker), this time, accounts and my data exists, it shows record of all my images and albums and accounts, but the images doesn't load. (i.e.- `postgres` could be accessed, but `library` couldn't)

```env
UPLOAD_LOCATION="D:\\immich\\library"
DB_DATA_LOCATION="C:\\Users\\<user>\\Downloads\\immich\\postgres"
```

I'm using the default configs of the [latest release (v1.118.2) as of October 20, 2024](https://github.com/immich-app/immich/releases/tag/v1.118.2)

Docker compose mounts are as follows:
```yml
volumes:
  - ${UPLOAD_LOCATION}:/usr/src/app/upload
  - ${DB_DATA_LOCATION}:/var/lib/postgresql/data
```

Note:
- Both `C:` and `D:` drives have NTFS partition.
- I reran `docker compose up -d` everytime I changed my config, and verified the mount points using Docker Desktop to be sure it's is using my intended path.
- Tried the re-installation and configurations multiple times, this behavior seems to be consistent.
- `D:` drive I refer to is an external Hard Disk (Single Partition, NTFS, ~1.2TB free)
- `C:` drive is internal SSD (Single Partition, NTFS, ~200GB free)

**Reply by dvishal485** [reply]
Okay after more trials I got some better insights. This **behavior is not consistent**. Sometimes external drive works, sometimes it doesn't. (as in immich works as expected with some non-zero probability `p`)

To make it work however, one can try these until the issue source is known and mitigated:
1. Try your luck, pray to god, restart container, reboot computer, hoping that everything shows right up.
2. Mount the drive to some path in internal drive using the Windows partition tool. This worked for me but it might be sheer luck. (This sounds super stupid though, but that's the observation)
    1. Open Windows Partition Tool.
    2. Right click on the external drive partition you want to use.
    3. Select "Change Drive Letters and Paths...".
    4. Select some empty folder in your `C:` drive (I used `C:\Users\<user>\Downloads\external_hdd`).
    5. Make sure your `.env` files uses relative path and not some hardcoded path.
    6. Run terminal from this directory, goto the immich folder and boot up docker).
    
    **Note:** Again, this might be coincidental, but everytime I ran immich using this path, it worked as expected.
